bundles/alpha.model/src/alpha/model/util/AffineFactorizer.xtend

```diff
@@ -0,0 +1,391 @@
+		// Name all the outputs from the expressions so we can track them.
+		// Note: this is saved as a map from the original expression to the named one.
+		val named = expressions.nameExpressionOutputs
+		
+		// Merge everything into one big expression, and decompose it.
+		// Since we need them again later, copy the named expressions first.
+		val decomposed = named.values.map[it.copy].mergeExpressions.decomposeExpression
+		val hExpression = decomposed.key
+		val qExpression = decomposed.value
+		
+		// Construct a map from each of the original expressions
+		// to the projection of Q which is associated with it.
+		// Wrap it in a HashMap to avoid recomputing values whenever they're accessed.
+		val projectionMap = new HashMap<ISLMultiAff, ISLMultiAff>(
+			expressions.toInvertedMap[expr | getDecompositionProjection(named.get(expr), qExpression)])
+		return hExpression -> projectionMap
+	}
+
+
+	////////////////////////////////////////////////////////////
+	// Merging Expressions Into One
+	////////////////////////////////////////////////////////////
+
+	/** 
+	 * Creates a copy of each input expression, but renames all the outputs so they can be tracked.
+	 * Returns a map from the original expressions to the named ones.
+	 */
+	def static nameExpressionOutputs(ISLMultiAff... expressions) {
+		val names = getNameGenerator("orig_out")
+		return expressions
+			.toInvertedMap[nameSingleExpressionOutputs(it, names)]
+			.toHashMap
+	}
+
+	/** Generates names containing the given prefix and an incrementing value, separated by an underscore. */
+	def private static getNameGenerator(String prefix) {
+		return (0..Integer.MAX_VALUE)
+			.map[prefix + "_" + it.toString()]
+			.iterator()
+	}
+
+	/** Gives each output of the provided expression a unique name, overwriting any existing names. */
+	def private static nameSingleExpressionOutputs(ISLMultiAff expression, Iterator<String> names) {
+		val outCount = expression.dim(ISLDimType.isl_dim_out)
+		val assignName = [ISLMultiAff expr, int outIndex | expr.setDimName(ISLDimType.isl_dim_out, outIndex, names.next)]
+		val merged = (0 ..< outCount).fold(expression.copy, assignName)
+		return merged
+	}
+	
+	/**
+	 * Wraps a map into a hash map so its values are only computed once.
+	 * Intended for use with Xtend's iterable/map extensions, which can recompute values
+	 * each time they're accessed due to lazy evaluation.
+	 */
+	def private static <K,V> toHashMap(Map<K,V> map) {
+		return new HashMap<K,V>(map)
+	}
+	
+	
+	////////////////////////////////////////////////////////////
+	// Expression to Matrix Conversion
+	////////////////////////////////////////////////////////////
+
+	/**
+	 * Converts an expression into a matrix of its parameters, input indexes, and constants.
+	 * The matrix is column-oriented (i.e., each affine expression is one column).
+	 * Throws an error if the expression uses division, as that may not work correctly.
+	 */
+	def static expressionToMatrix(ISLMultiAff expression) {
```

Comment:
Could we add a comment to explain/remind how this is different from this function:
https://github.com/CSU-CS-Melange/alpha-language/blob/affine-factorization/bundles/alpha.model/src/alpha/model/util/AffineFunctionOperations.java#L69

They both take an ISLMultiAff and return an ISLMatrix. I remember you mentioned that what you need here is a matrix with at most one column of zeros. This is needed for the subsequent input the hermite decomposition, right?